### PR TITLE
Add a NuGet package for command line Code Metrics

### LINKED ISSUE #1756: Shipping story for Metrics.exe
https://github.com/dotnet/roslyn-analyzers/pull/1754 ports Metrics.exe that shipped from VS2010-VS2015 for code metrics report generation. The PR does not package or publish the exe. For now, the customers have to build the repo locally and produce the exe. If this is not sufficient and we have enough customer requests, we need to sign/package/publish this exe.

- [ ] Localization
- [ ] Unit tests
- [ ] Signing
- [ ] Packaging
- [ ] Publishing

### PR DESCRIPTION
**Pull request by mavasani**
Fixes #1756
See https://docs.microsoft.com/en-us/visualstudio/code-quality/how-to-generate-code-metrics-data?view=vs-2017#command-line-code-metrics for background

1. Users can now execute Metrics.exe by installing the `Microsoft.CodeMetrics` NuGet package and then executing `msbuild /t:Metrics` on their project file.
2. Metrics can be executed in legacy mode by specifying `/p:LEGACY_CODE_METRICS_MODE=true` on the MSBuild command line.
3. Additionally, users can override the output file by specifying `/p:MetricsOutputFile=<filename>` on the MSBuild command line.

Verified that the locally generated Microsoft.CodeMetrics NuGet package can be installed on a managed project and above functionality works.

------

**Example 1:**
```
C:\Users\mavasani\source\repos\ClassLibrary3\ClassLibrary3>msbuild /t:Metrics
Microsoft (R) Build Engine version 16.0.360-preview+g9781d96883 for .NET Framework
Copyright (C) Microsoft Corporation. All rights reserved.

Build started 1/22/2019 4:29:57 PM.
Project "C:\Users\mavasani\source\repos\ClassLibrary3\ClassLibrary3\ClassLibrary3.csproj" on node 1 (Metrics target(s))
.
Metrics:
  C:\Users\mavasani\source\repos\ClassLibrary3\packages\Microsoft.CodeMetrics.2.6.4-ci\build\\..\Metrics\Metrics.exe /p
  roject:C:\Users\mavasani\source\repos\ClassLibrary3\ClassLibrary3\ClassLibrary3.csproj /out:ClassLibrary3.Metrics.xml
  Loading ClassLibrary3.csproj...
  Computing code metrics for ClassLibrary3.csproj...
  Writing output to 'ClassLibrary3.Metrics.xml'...
  Completed Successfully.
Done Building Project "C:\Users\mavasani\source\repos\ClassLibrary3\ClassLibrary3\ClassLibrary3.csproj" (Metrics target
(s)).


Build succeeded.
    0 Warning(s)
    0 Error(s)
```

------

**Example 2 (Legacy mode with overridden output file):**
```
C:\Users\mavasani\source\repos\ClassLibrary3\ClassLibrary3>msbuild /t:Metrics /p:LEGACY_CODE_METRICS_MODE=true /p:MetricsOutputFile="Legacy.xml"
Microsoft (R) Build Engine version 16.0.360-preview+g9781d96883 for .NET Framework
Copyright (C) Microsoft Corporation. All rights reserved.

Build started 1/22/2019 4:31:00 PM.
The "MetricsOutputFile" property is a global property, and cannot be modified.
Project "C:\Users\mavasani\source\repos\ClassLibrary3\ClassLibrary3\ClassLibrary3.csproj" on node 1 (Metrics target(s))
.
Metrics:
  C:\Users\mavasani\source\repos\ClassLibrary3\packages\Microsoft.CodeMetrics.2.6.4-ci\build\\..\Metrics.Legacy\Metrics
  .Legacy.exe /project:C:\Users\mavasani\source\repos\ClassLibrary3\ClassLibrary3\ClassLibrary3.csproj /out:Legacy.xml
  Loading ClassLibrary3.csproj...
  Computing code metrics for ClassLibrary3.csproj...
  Writing output to 'Legacy.xml'...
  Completed Successfully.
Done Building Project "C:\Users\mavasani\source\repos\ClassLibrary3\ClassLibrary3\ClassLibrary3.csproj" (Metrics target
(s)).


Build succeeded.
    0 Warning(s)
    0 Error(s)
```

### CODE REVIEW

**Comment by mavasani** [comment]
Tagging @gewarren - if/when this goes in, we would need to update the documentation on Code Metrics (I will file a separate documentation bug once that happens).

Also FYI for @dotnet/roslyn-analysis - I have got numerous requests internally and externally for this NuGet package, so it would be good to take this forward.

File: nuget/Microsoft.CodeMetrics/Microsoft.CodeMetrics.targets [hunk 1]
```diff
@@ -0,0 +1,19 @@
+<!-- Copyright (c)  Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information. -->
+<Project>
+  <PropertyGroup>
+    <MetricsExeFolder>$(MSBuildThisFileDirectory)\..\Metrics</MetricsExeFolder>
+    <MetricsExeName>Metrics</MetricsExeName>
+    <MetricsOutputFile Condition="'$(MetricsOutputFile)' == ''">$(MSBuildProjectName).Metrics.xml</MetricsOutputFile>
```
Comment (genlu) [inline nuget/Microsoft.CodeMetrics/Microsoft.CodeMetrics.targets]: Is this required for legacy mode too?

Comment (genlu) [reply nuget/Microsoft.CodeMetrics/Microsoft.CodeMetrics.targets]: nit: maybe use `Choose-When` instead

File: eng/GenerateAnalyzerNuspec.targets [hunk 2]
```diff
@@ -77,6 +77,6 @@
       <_CscToolPath Condition="!HasTrailingSlash('$(_CscToolPath)')">$(_CscToolPath)\</_CscToolPath>
     </PropertyGroup>
 
-    <Exec Command='"$(_CscToolPath)csi.exe" "$(RepoRoot)eng\GenerateAnalyzerNuspec.csx" "$(NuspecFile)" "$(AssetsDir)\" "$(MSBuildProjectDirectory)" "$(Configuration)" "$(TargetFramework)" "@(_NuspecMetadata)" "@(AnalyzerNupkgFile)" "@(AnalyzerNupkgAssembly)" "@(AnalyzerNupkgDependency)" "$(_GeneratedRulesetsDir)" "@(AnalyzerLegacyRuleset)"' />
+    <Exec Command='"$(_CscToolPath)csi.exe" "$(RepoRoot)eng\GenerateAnalyzerNuspec.csx" "$(NuspecFile)" "$(AssetsDir)\" "$(MSBuildProjectDirectory)" "$(Configuration)" "$(TargetFramework)" "@(_NuspecMetadata)" "@(AnalyzerNupkgFile)" "@(AnalyzerNupkgFolder)" "@(AnalyzerNupkgAssembly)" "@(AnalyzerNupkgDependency)" "$(_GeneratedRulesetsDir)" "@(AnalyzerLegacyRuleset)" "$(ArtifactsBinDir)\"' />
```
Comment (genlu) [inline eng/GenerateAnalyzerNuspec.targets]: It's nice that this repo still uses csx for infrastructure :)

File: nuget/Microsoft.CodeMetrics/Microsoft.CodeMetrics.targets (hunk 1 above)
Comment (mavasani) [reply nuget/Microsoft.CodeMetrics/Microsoft.CodeMetrics.targets]: I'll stick with the current approach as the properties within the legacy property group just need to append `.Legacy`

File: eng/GenerateAnalyzerNuspec.targets (hunk 2 above)
Comment (mavasani) [reply eng/GenerateAnalyzerNuspec.targets]: Yes, Tomas' work mostly :-)

**Comment by mavasani** [comment]
NuGet package to try out is available here: https://dotnet.myget.org/feed/roslyn-analyzers/package/nuget/Microsoft.CodeMetrics

**Comment by mavasani** [comment]
NuGet package has been renamed and published to NuGet.org: https://www.nuget.org/packages/Microsoft.CodeAnalysis.Metrics/

**Comment by dacezarina** [comment]
Can you please make this package work also with dotnet msbuild on Mac or Linux? So it would be ideal to generate code metrics for non-windows environments.

**Comment by genlu** [comment]
@dacezarina Could you please file an issue for your request? Thanks!

**Comment by dacezarina** [comment]
Created an issue for this https://github.com/dotnet/roslyn-analyzers/issues/2402

**Pull request merged**